scripts/runnable_recipe_ray_vllm_weight_sync.py

```diff
@@ -1582,6 +2031,204 @@ def close(self):
             self.logger.close()
 
 
+class vLLMHFLocalWeightUpdater(LocalWeightUpdaterBase):
+    def __init__(self, master_address, master_port, model_metadata):
+        print(f"{master_address=}, {master_port=}")
+        self.master_address = master_address
+        self.master_port = master_port
+        self.model_metadata = model_metadata
+        self.initialized_group = None
+
+    def _get_server_weights(self):
+        return None
+
+    def _get_local_weights(self):
+        # We don't implement this because we let vLLM's update_weights API handle everything for now
+        return None
+    
+    def _maybe_map_weights(self, server_weights, local_weights):
+        # vLLM update_weights function handles the mapping from huggingface
+        # so we don't implement this for now
+        return None
+    
+    def _update_local_weights(self, local_weights, mapped_weights):
+        llm = self.collector.llm
+        if self.initialized_group is None:
+            weight_sync_world_size = llm.llm_engine.parallel_config.tensor_parallel_size + 1
+            llm.collective_rpc(
+                "init_weight_update_group",
+                args=(self.master_address, self.master_port, 1, weight_sync_world_size)
+            )
+            self.initialized_group = True
+        
+        for k, (dtype, shape) in self.model_metadata.items():
+            llm.collective_rpc(
+                "update_weight",
+                args=(k, dtype, shape)
+            )
+    
+        llm.collective_rpc("update_policy_version")
+        print("done local update_weight")
+
+import threading
+
+class ReadWriteLock:
```

Comment:
NEVER write these classes yourself. Even the simplest cases like this are more tricky. Your function is similar to this [here](https://stackoverflow.com/questions/22079500/one-producer-thread-and-multiple-non-blocking-simultaneous-consumer-threads-mro) (at the bottom) but you can see that they have more logic. 

Here, I would use it from this package: https://github.com/elarivie/pyReaderWriterLock

Reply:
Also, using the package gives you the ability to set a timeout, which is always useful in these situations